### PR TITLE
Baba/aux bms functional

### PR DESCRIPTION
**Pull request by badeniran**
Functional code for the Aux BMS. Will still need to be changed once Dan gets USB header.

### CODE REVIEW

**Comment by bill-luu** [comment]
Is this a continuation of the last PR? If so, can we close that one and then review this?

**Comment by badeniran** [comment]
@bill-luu Yeah sounds like a good plan

File: EpsilonAuxBMS/Src/AuxBMS.c [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+        // Data buffer
+        uint8_t rxBuff[2] = {0};
+        // Set Chip Select to be low
+        HAL_GPIO_WritePin(ADC_nCS_GPIO_Port, ADC_nCS_Pin, GPIO_PIN_RESET);
+
+        // Start SCLK and receive input
+        if (HAL_SPI_Receive(&hspi3, rxBuff, size, SPI_TIMEOUT) == HAL_OK)
+        {
+            // ADC sends 4 (could be 3, will get second opinion) zeros before sending data,
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/AuxBMS.c]: Do you still need that second opinion?

File: EpsilonAuxBMS/Src/AuxBMS.c [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+            //so can ignore top 4 bits of rxBuff[0] and bottom 2 bits of rxBuff[1]
+            uint16_t upperBits = (uint16_t)(rxBuff[1] << 6);
+            uint16_t lowerBits = (uint16_t)(rxBuff[0] & 0xF6); // Don't care about bottom 2 bits
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/AuxBMS.c]: This comment is already covered by the earlier comment

File: EpsilonAuxBMS/Src/AuxBMS.c [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+        }
+        else
+        {
+            voltage = 0;
+        }
+
+        // Set Chip Select to be high
+        HAL_GPIO_WritePin(ADC_nCS_GPIO_Port, ADC_nCS_Pin, GPIO_PIN_SET);
+
+
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/AuxBMS.c]: Excess whitespace

File: EpsilonAuxBMS/Src/AuxBMS.c [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+        {
+            *current_sense = HAL_ADC_GetValue(&hadc1);
+            int diff = *current_sense - *prev_sense;
+
+            if (diff < 0)
+            {
+                diff *= -1;
+            }
+
+            if (diff > 50)
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/AuxBMS.c]: Why is there this wait between senses?

File: EpsilonAuxBMS/Src/AuxBMS.c [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+    }
+}
+
+void readCurrent(int* counter, uint32_t* current_sense, uint32_t* prev_sense)
+{
+    // Enable current amplifier. Allow time to settle and take multiple valid concurrent readings
+    // to make sure it isn't changing
+    HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_SET);
+
+    while (*counter < 3)
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/AuxBMS.c]: How come we only need 3 good adc reads?

File: EpsilonAuxBMS/Src/AuxCAN.c [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+        // Toggle Red LED for every heartbeat sent
+        HAL_GPIO_TogglePin(RED_LED_GPIO_Port, RED_LED_Pin);
+
+        // Set CAN message address
+        hcan1.pTxMsg->StdId = AUX_HEARTBEAT_STDID;
+        // Always 1
+        hcan1.pTxMsg->Data[0] = 1;
+        // Send CAN message
+        HAL_CAN_Transmit_IT(&hcan1);
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/AuxCAN.c]: Can you check if this function returns some kind of status? If it does, let's toggle the red pin every time it returns HAL_OK

File: EpsilonAuxBMS/Src/AuxCAN.c [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        osDelayUntil(&prevWakeTime, AUX_STATUS_FREQ);
+
+        if (osMutexWait(canHandleMutex, 0) != osOK)
+        {
+            continue;
+        }
+
+        // Toggle blue LED for every CAN message sent
+        HAL_GPIO_TogglePin(BLU_LED_GPIO_Port, BLU_LED_Pin);
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/AuxCAN.c]: Similar to the comment on the RED LED, check if it gives a HAL_OK

File: EpsilonAuxBMS/Src/AuxCAN.c [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    }
+}
+
+// Reimplement weak definition in stm32f4xx_hal_can.c
+void HAL_CAN_RxCpltCallback(CAN_HandleTypeDef* hcan)
+{
+    CanRxMsgTypeDef* msg = hcan->pRxMsg;
+
+    // Toggle green LED for every CAN message received
+    HAL_GPIO_TogglePin(GRN_LED_GPIO_Port, GRN_LED_Pin);
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/AuxCAN.c]: Same thing with the other LEDs

File: EpsilonAuxBMS/Inc/AuxBMS.h [hunk 9]
```diff
@@ -0,0 +1,92 @@
+#pragma once
```
Comment (bill-luu) [inline EpsilonAuxBMS/Inc/AuxBMS.h]: This file is included by both `AuxBMS.h` & `AuxCAN.h`, that's kinda weird. Let's just have two different AuxBMS & AuxCAN headers.

File: EpsilonAuxBMS/Src/AuxCAN.c [hunk 10]
```diff
@@ -0,0 +1,87 @@
+#include "cmsis_os.h"
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/AuxCAN.c]: I'd consider renaming this file to something like CanSending.

File: EpsilonAuxBMS/Src/AuxBMS.c [hunk 11]
```diff
@@ -0,0 +1,340 @@
+#include "cmsis_os.h"
```
Comment (bill-luu) [inline EpsilonAuxBMS/Src/AuxBMS.c]: This file is getting to be pretty busy, consider breaking it up into different files (One for each Task).

File: EpsilonAuxBMS/Src/AuxBMS.c (hunk 4 above)
Comment (badeniran) [reply EpsilonAuxBMS/Src/AuxBMS.c]: Dan had said it would be a good idea to give the current sensor a bit of time before reading it.

File: EpsilonAuxBMS/Src/AuxBMS.c (hunk 5 above)
Comment (badeniran) [reply EpsilonAuxBMS/Src/AuxBMS.c]: Dan had said we should get a few consistent reads so I just sent with 3. It could be more

File: EpsilonAuxBMS/Src/AuxCAN.c (hunk 10 above)
Comment (badeniran) [reply EpsilonAuxBMS/Src/AuxCAN.c]: There's the receiving callback in there so would CanHandler or something be suitable?

Comment (bill-luu) [reply EpsilonAuxBMS/Src/AuxCAN.c]: Yeah!

**Comment by bill-luu** [comment]
Format

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+    // If current is high for some reason or sense pin is high something is wrong
+    if (pwr_current > CURRENT_LOWER_THRESHOLD * current_multiplier || !sense_pin)
+    {
+        return 0;
+    }
+
+    return 1;
+}
+
+uint32_t readCurrent(void)
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: This function has global scope, it should have a better name.

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+
+            default:
+                state = FIRST_CHECK;
+        }
+
+        osMutexRelease(auxStatusMutex);
+    }
+}
+
+int checkContactor(uint16_t pin, GPIO_TypeDef* port, int current_multiplier)
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: This function has global scope, it should have a better name.

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+{
+    float pwr_voltage = 0.0;
+    float pwr_current = 0.0;
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: naming convnetion is capCase.

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+uint32_t readCurrent(void)
+{
+    // Enable current amplifier. Allow time to settle and take multiple valid concurrent readings
+    // to make sure it isn't changing
+    HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_SET);
+    uint32_t sense = 0;
+    uint32_t prev_sense = 0;
+    int counter = 0;
+
+    while (counter < 3)
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: This is better represente as a for loop.

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+            {
+                counter++;
+            }
+        }
+        else
+        {
+            sense = 0xFFFFFFFF;
+            break;
+        }
+    }
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: Can you explain how this peripheral works as a comment at the top of the func?

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+
+static const uint32_t AUX_SET_CONTACTOR_FREQ = 1000; // Every 1 second
+static const float CURRENT_SENSE_RESISTOR  = 0.001; //actually 1 mOhm, but 1 Ohm for testing
+static const int GAIN  = 250;
+static const uint32_t SENSE_SETUP_TIME  = 50; // Setup time for current_sense
+static const float CURRENT_LOWER_THRESHOLD  = 0.15; // Lower current threshold
+static const uint32_t ADC_POLL_TIMEOUT  = 10;
+static const int MAX_ATTEMPTS  = 2; // max attempts for trying to turn on a contactor
+
+typedef enum State {FIRST_CHECK, COMMON_CONTACTOR_ON, CHARGE_CONTACTOR_ON, DISCHARGE_CONTACTOR_ON,
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: This is a global variable, it should have a better name, and also this looks kinda messy, would be better to have it on one line each.

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+    for (;;)
+    {
+        osDelayUntil(&prevWakeTime, AUX_SET_CONTACTOR_FREQ);
+
+        if (osMutexWait(auxStatusMutex, 0) != osOK)
+        {
+            continue;
+        }
+
+        if (attemptCount > MAX_ATTEMPTS) // If it's greater than a certain number of attempts, report to CAN that something is wrong
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: Having the number of attempts intertwined into this function is pretty hard to follow. It'd be better to just have a `setContactors()` which could return an error, and for every error the function gets run again.

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+static const uint32_t SENSE_SETUP_TIME  = 50; // Setup time for current_sense
+static const float CURRENT_LOWER_THRESHOLD  = 0.15; // Lower current threshold
+static const uint32_t ADC_POLL_TIMEOUT  = 10;
+static const int MAX_ATTEMPTS  = 2; // max attempts for trying to turn on a contactor
+
+typedef enum State {FIRST_CHECK, COMMON_CONTACTOR_ON, CHARGE_CONTACTOR_ON, DISCHARGE_CONTACTOR_ON,
+                    DONE, BLOCKED
+                   } State;
+
+void setContactorsTask(void const* arg)
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: Since this is a very specific process, can you provide a comment at hte top of how this all is supposed to work?

File: EpsilonAuxBMS/Src/UpdateChargeStatusTask.c [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+            orionStatus.batteryVoltagesInRange = 0;
+            // Turn off High voltage enable and discharge contactor
+            HAL_GPIO_WritePin(HV_ENABLE_GPIO_Port, HV_ENABLE_Pin, GPIO_PIN_RESET);
+            HAL_GPIO_WritePin(CONTACTOR_ENABLE3_GPIO_Port, CONTACTOR_ENABLE3_Pin, GPIO_PIN_RESET);
+            auxStatus.dischargeContactorState = 0;
+        }
+        else
+        {
+            orionStatus.batteryVoltagesInRange = 1;
+        }
```
[the remark this answers is not in the document]
Comment (badeniran) [reply EpsilonAuxBMS/Src/UpdateChargeStatusTask.c]: We have 4, since there's two different cells. Cell with highest voltage (if too high or in range) and Cell with lowest voltage (if too low or in range)

File: EpsilonAuxBMS/Src/UpdateChargeStatusTask.c [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+    // For concurrency with setContactorsTask()
+    osMutexId* auxStatusMutex = (osMutexId*) arg;
+    // One time osDelayUntil initialization
+    uint32_t prevWakeTime = osKernelSysTick();
+
+    for (;;)
+    {
+        osDelayUntil(&prevWakeTime, CHARGE_STATUS_UPDATE_FREQ);
+
+        if (osMutexWait(auxStatusMutex, 0) != osOK)
```
[the remark this answers is not in the document]
Comment (badeniran) [reply EpsilonAuxBMS/Src/UpdateChargeStatusTask.c]: Are you referring to the pin writes going on in other tasks? I'm not too sure as to what you mean here

File: EpsilonAuxBMS/Src/SetContactorsTask.c (hunk 16 above)
Comment (badeniran) [reply EpsilonAuxBMS/Src/SetContactorsTask.c]: I should explain how the ADC works or the current amplifier?

Comment (enochtsang) [reply EpsilonAuxBMS/Src/SetContactorsTask.c]: This function does some pretty weird stuff, why is there multiple attempts and multiple reads?

File: EpsilonAuxBMS/Src/UpdateChargeStatusTask.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (enochtsang) [reply EpsilonAuxBMS/Src/UpdateChargeStatusTask.c]: Right, I remember now. That means there's a logical error though! If the maxCellVoltage is too high, and the minCellVoltage is ok, then `orionStatus.batteryVoltagesInRange` will be 1, which is not true since it's too high. Also this is the second time I've been confused at this, maybe a comment is necessary haha, I don't think I'll be the first person to miss this. @

File: EpsilonAuxBMS/Src/ReportAuxStatusToCanTask.c [hunk 22]
```diff
@@ -23,16 +23,14 @@ void reportAuxStatusToCanTask(void const* arg)
...
-        hcan1.pTxMsg->Data[0] += auxStatus.commonContactorState * 0x1;
-        hcan1.pTxMsg->Data[0] += auxStatus.chargeContactorState * 0x2;
-        hcan1.pTxMsg->Data[0] += auxStatus.dischargeContactorState * 0x4;
-        hcan1.pTxMsg->Data[0] |= auxStatus.auxVoltage * 0x8;
-        // Initialize to 0 so there's no garbage in [2] to [7]
-        hcan1.pTxMsg->Data[1] = 0;
-        hcan1.pTxMsg->Data[1] += auxStatus.strobeBmsLight * 0x1;
-        hcan1.pTxMsg->Data[1] += auxStatus.allowCharge * 0x2;
-        hcan1.pTxMsg->Data[1] += auxStatus.contactorError * 0x4;
+        hcan1.pTxMsg->Data[0] += auxStatus.commonContactorState;
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/ReportAuxStatusToCanTask.c]: these should all be ORd together. The semantics of `+=` for bitwise operations is wrong.

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 23]
```diff
@@ -1,121 +1,118 @@
 #include "SetContactorsTask.h"
+/*
+  This task allows for the turnin gon of the common, charge, and discharge contactors.
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: `turnin gon` typo

File: EpsilonAuxBMS/Src/main.c [hunk 24]
```diff
@@ -0,0 +1,630 @@
...
+/* USER CODE BEGIN 4 */
+
+// Reimplement weak definition in stm32f4xx_hal_can.c
+void HAL_CAN_RxCpltCallback(CAN_HandleTypeDef* hcan)
+{
+    CanRxMsgTypeDef* msg = hcan->pRxMsg;
+
+    if (msg->StdId == ORION_MAX_MIN_VOLTAGES_STDID)
+    {
+        orionStatus.maxCellVoltage = (uint16_t)msg->Data[6]; // Max Cell voltage
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/main.c]: orionStatus needs mutexes.

File: EpsilonAuxBMS/Src/UpdateChargeAllowanceTask.c [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+static const float MIN_CELL_VOLTAGE = 3.2;
+typedef enum ChargeAllowanceState {VOLTAGE_IN_RANGE,
+                                   OUT_OF_RANGE
+                                  } ChargeAllowanceState;
+
+void updateChargeAllowanceTask(void const* arg)
+{
+    // One time osDelayUntil initialization
+    uint32_t prevWakeTime = osKernelSysTick();
+    ChargeAllowanceState maxCellVoltageChargeAllowanceState = VOLTAGE_IN_RANGE;
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/UpdateChargeAllowanceTask.c]: The logic for this file can be simpler, try something like this.

```C
void updateChargeAllowanceTask(void const* arg)
{
    // One time osDelayUntil initialization
    uint32_t prevWakeTime = osKernelSysTick();

    for (;;)
    {
        osDelayUntil(&prevWakeTime, CHARGE_ALLOWANCE_UPDATE_FREQ);

        /* Orion BMS sends voltages relating to two DIFFERENT cells: the cell with the highest voltage
           and the cell with the lowest voltage. Must check if the voltage of the cell with the highest
           voltage is too high or in range and if the cell with the lowest voltage is too low or in range.
        */
        uint8_t voltageInRange = 1;
        uint8_t allowCharge = 1;
        uint8_t allowDischarge = 1;

        if ((double)orionStatus.maxCellVoltage > MAX_CELL_VOLTAGE * 0.8) // Will have the cutoff to be 20% below
        {
            // Turn off charge contactor
            HAL_GPIO_WritePin(CONTACTOR_ENABLE2_GPIO_Port, CONTACTOR_ENABLE2_Pin, GPIO_PIN_RESET);
            allowCharge = 0;
            voltageInRange = 0;
        }

        if ((double)orionStatus.minCellVoltage < MIN_CELL_VOLTAGE * 1.2) // Will have the cutoff to be 20% above
        {
            // Turn off High voltage enable and discharge contactor
            HAL_GPIO_WritePin(HV_ENABLE_GPIO_Port, HV_ENABLE_Pin, GPIO_PIN_RESET);
            HAL_GPIO_WritePin(CONTACTOR_ENABLE3_GPIO_Port, CONTACTOR_ENABLE3_Pin, GPIO_PIN_RESET);
            allowDischarge = 0;
            voltageInRange = 0;
        }

        // Setting the allowance of charge and charge/discharge contactor state for auxStatus
        if (osMutexWait(auxStatus.auxStatusMutex, 0) != osOK)
        {
            continue;
        }

        auxStatus.allowCharge = allowCharge;
        auxStatus.allowDischarge = allowDischarge;
        orionStatus.batteryVoltagesInRange = voltageInRange;

        osMutexRelease(auxStatus.auxStatusMutex);
    }
}
```

Overall I would say this file should just update some internal state. And let `setContactorTask()` actually physically deal with the contactors.

Also a new question in my head, you used the 20% safety factors, did electrical say this was ok? Because of how voltage works, they're usually in a curve, so 20% voltage isn't actually 20% battery usage, it will probably be more like 40% battery usage. Some example voltage curves (SOC curves), https://chargedevs.com/wp-content/uploads/2013/09/SOC-Measurments-figure11.png.

Also what task is supposed to be controlling the `HV_ENABLE_GPIO`? You turn it off here, but is it possible that it will get turned on again somewhere else?

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 26]
```diff
@@ -0,0 +1,284 @@
...
+
+                if (updateContactorState(contactorState, DISCHARGE))
+                {
+                    state = DONE;
+                }
+
+                break;
+
+            case DONE: ; // Put this here because compiler kept complaining about a declaration not being a statement
+                // could potentially cause problems.
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: This warning is legit. What is happening here is the the scope that's in this case is leaking.

`uint8_t common` and others have their scope in the whole function right now. What should be done, is scoping.

```C
case DONE: {
    uint8_t common;
    ...
    break;
}
```

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 27]
```diff
@@ -0,0 +1,284 @@
...
+                uint8_t common = !HAL_GPIO_ReadPin(SENSE1_GPIO_Port, SENSE1_Pin);
+                uint8_t charge = !HAL_GPIO_ReadPin(SENSE2_GPIO_Port, SENSE2_Pin);
+                uint8_t discharge = !HAL_GPIO_ReadPin(SENSE3_GPIO_Port, SENSE3_Pin);
+
+                if (!common && !charge && !discharge) // None of the contactors are enabled
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: For safety, we should just use `||` here instead of `&&`, reset everything is any of them are not set and start from the beginning. Then alot of the code below is not necessary.

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 28]
```diff
@@ -0,0 +1,284 @@
...
+                        state = DISCHARGE_CONTACTOR_CHECK;
+                    }
+                }
+
+                break;
+
+            case BLOCKED:
+                if (orionStatus.gpioOk && orionStatus.batteryVoltagesInRange)
+                {
+                    state = prevState;
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: We should reset and start from the first state imo.

File: EpsilonAuxBMS/Src/SetContactorsTask.c [hunk 29]
```diff
@@ -0,0 +1,284 @@
...
+            }
+        }
+        else
+        {
+            sense = 0xFFFFFFFF;
+            break;
+        }
+    }
+
+    HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_RESET);
```
Comment (enochtsang) [inline EpsilonAuxBMS/Src/SetContactorsTask.c]: Because of the osDelay for hte sense setup time, the ADC is going to be held for alot longer than it needs to. This pin set and reset should be done once for everytime the value, not once for the whole function.

File: EpsilonAuxBMS/Src/main.c [hunk 30]
```diff
@@ -0,0 +1,630 @@
...
+    osThreadDef(setContactorTask, setContactorsTask, osPriorityNormal, 1, configMINIMAL_STACK_SIZE);
+    setContactorsTaskHandle = osThreadCreate(osThread(setContactorTask), NULL);
+    // Setup task to read aux voltage
+    osThreadDef(readVoltageTask, readAuxVoltageTask, osPriorityNormal, 1, configMINIMAL_STACK_SIZE);
+    readAuxVoltageTaskHandle = osThreadCreate(osThread(readVoltageTask), NULL);
+    // Setup task to report aux status to CAN
+    osThreadDef(auxStatusCanTask, reportAuxStatusToCanTask, osPriorityNormal, 1, configMINIMAL_STACK_SIZE);
+    reportAuxStatusToCanTaskHandle = osThreadCreate(osThread(auxStatusCanTask), canHandleMutex);
+    // Setup task to report hearbeat to CAN
+    osThreadDef(heartbeatTask, reportHeartbeatToCanTask, osPriorityNormal, 1, configMINIMAL_STACK_SIZE);
```
[the remark this answers is not in the document]
Comment (badeniran) [reply EpsilonAuxBMS/Src/main.c]: When you're saying consistency, are you talking about consistency between the first and second arguments or all the first arguments?

File: EpsilonAuxBMS/Src/UpdateChargeAllowanceTask.c (hunk 25 above)
Comment (badeniran) [reply EpsilonAuxBMS/Src/UpdateChargeAllowanceTask.c]: The 20% was directly from Dan and Logan. The reason I'm also messing with the contactors here is because they said there needs to be another, much faster task, that can shut everything down ASAP. SetContactors is a much slower task. HV_ENABLE is turned on at the very end of setContactors. Could possibly make another task just for that, but it's one line so I didn't think it was necessary

File: EpsilonAuxBMS/Src/SetContactorsTask.c (hunk 29 above)
Comment (badeniran) [reply EpsilonAuxBMS/Src/SetContactorsTask.c]: I can't reset that pin because if I do, it defeats the purpose of the function. By doing the setting and resetting, I'm turning the current amplifier on and off. So we're trying to get good reads after turning it on. Then once we get the consistency, it can go off.

File: EpsilonAuxBMS/Src/ReadAuxVoltageTask.c [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+            // Set Chip Select to be high. Placed here to keep pin low as short as possible
+            HAL_GPIO_WritePin(ADC_nCS_GPIO_Port, ADC_nCS_Pin, GPIO_PIN_SET);
+            // ADC sends 10 bits of data with MSB being sent first.
+            // ADC sends 4 zeros before sending data, so can ignore top 4 bits of rxBuff[1] (only want bottom 4 bits).
+            // Want bit 3 of rxBuff[1] to become bit 9 so left shift
+            uint16_t upperBits = (uint16_t)(rxBuff[1] << 6);
+            // Bottom 2 bits of rxBuff[0] will be meaningless so ignore as well (only want top 6 bits).
+            // Right shift lowerBits to remove bottom 2 bits and line up bit 7 of rxBuff[0] to bit 5
+            uint16_t lowerBits = (uint16_t)((rxBuff[0]) >> 2);
+            spiVoltage =  0x03FF & (upperBits | lowerBits);
```
[the remark this answers is not in the document]
Comment (badeniran) [reply EpsilonAuxBMS/Src/ReadAuxVoltageTask.c]: The HAl_SPI read function on accepts 8bit buffers

File: EpsilonAuxBMS/Src/UpdateChargeAllowanceTask.c [hunk 32]
```diff
@@ -24,15 +25,15 @@ void updateChargeAllowanceTask(void const* arg)
         uint8_t allowCharge = 1;
         uint8_t allowDischarge = 1;
 
-        if ((float)orionStatus.maxCellVoltage > MAX_CELL_VOLTAGE * 0.8) // Will have the cutoff to be 20% below
+        if (DEFAULT_VOLTAGE_UNITS * (float)orionStatus.maxCellVoltage > MAX_CELL_VOLTAGE * 0.8) // Will have the cutoff to be 20% below
```
[the remark this answers is not in the document]
Comment (badeniran) [reply EpsilonAuxBMS/Src/UpdateChargeAllowanceTask.c]: The datatype of orionStatus voltages is uint16_t, so I was just type casting for doing the multiplication. No sure if I need to or not. I can take it out and we'll see what happens

**Comment by bill-luu** [comment]
Context for new changes: #70
